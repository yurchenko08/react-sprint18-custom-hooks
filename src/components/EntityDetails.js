import { useState, useEffect } from "react";

function EntityDetails({ fetchMethod, params, propsToDisplay }) {
  const [entity, setEntity] = useState(null);

  useEffect(() => {
    setEntity(null);
    fetchMethod(params).then((res) => {
      setEntity(res);
      console.log("User: ", res);
    });
  }, [fetchMethod, params]);

  return entity ? (
    <div className="center">
      <h3>{entity.name} Details:</h3>
      {Object.entries(propsToDisplay).map((entry) => (
        <div key={entry[1]}>
          {" "}
          {entry[1]}: <strong>{entity[entry[0]]}</strong>
        </div>
      ))}
    </div>
  ) : (
    <p className="center">Loading...</p>
  );
}

export default EntityDetails;
