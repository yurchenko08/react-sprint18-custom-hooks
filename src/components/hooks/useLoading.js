import { useState, useEffect } from 'react';
export const useLoading = (fetchMethod, params) => {
  const [entity, setEntity] = useState(null);

  useEffect(() => {
    fetchMethod(params).then((res) => {
      setEntity(res);
      console.log('Users: ', res);
    });
  }, [fetchMethod, params]);

  return entity;
};
