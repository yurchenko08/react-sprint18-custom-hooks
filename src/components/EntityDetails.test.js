import renderer from "react-test-renderer";
import EntityDetails from "../components/EntityDetails";
import TestRenderer from "react-test-renderer";
import { setImmediate } from "timers";
import * as useLoadingContext from "./hooks/useLoading";
import { render } from "@testing-library/react";

const { act } = TestRenderer;

describe("EntityDetails", () => {
  it("renders correctly with data", async () => {
    const wrapper = renderer.create(
      <EntityDetails
        fetchMethod={() =>
          new Promise((resolve) =>
            resolve({ id: 101, name: "John", email: "john@gmail.com" })
          )
        }
        propsToDisplay={{ name: "Name" }}
      />
    );
    await act(() => new Promise(setImmediate));

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it("renders loading without data", async () => {
    const wrapper = renderer.create(
      <EntityDetails
        fetchMethod={() =>
          new Promise((resolve) =>
            resolve({ id: 101, name: "John", email: "john@gmail.com" })
          )
        }
        propsToDisplay={{ name: "Name" }}
      />
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it("checks if useLoading was called", () => {
    const mySpy = jest
      .spyOn(useLoadingContext, "useLoading")
      .mockImplementation(() => []);
    render(
      <EntityDetails
        fetchMethod={() => new Promise(() => [])}
        propsToDisplay={{}}
      />
    );

    expect(mySpy).toHaveBeenCalledTimes(1);
  });
});
