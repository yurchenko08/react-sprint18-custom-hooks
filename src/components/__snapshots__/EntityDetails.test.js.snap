// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityDetails renders correctly with data 1`] = `
<div
  className="center"
>
  <h3>
    John
     Details:
  </h3>
  <div>
     
    Name
    : 
    <strong>
      John
    </strong>
  </div>
</div>
`;

exports[`EntityDetails renders loading without data 1`] = `
<p
  className="center"
>
  Loading...
</p>
`;
